massa-execution-worker/src/active_history.rs

```diff
@@ -0,0 +1,106 @@
+use massa_execution_exports::ExecutionOutput;
+use massa_hash::Hash;
+use massa_ledger_exports::{
+    LedgerEntry, LedgerEntryUpdate, SetOrDelete, SetOrKeep, SetUpdateOrDelete,
+};
+use massa_models::{Address, Amount};
+use std::collections::VecDeque;
+
+#[derive(Default)]
+pub(crate) struct ActiveHistory(pub VecDeque<ExecutionOutput>);
+
+/// Result of a lazy, active history search
+pub enum HistorySearchResult<T> {
+    Found(T),
+    NotFound,
+    Deleted,
+}
+
+impl ActiveHistory {
+    /// Lazily query (from end to beginning) the active balance of an address after a given index.
+    ///
+    /// Returns a `HistorySearchResult`.
+    pub fn fetch_active_history_balance(
+        &self,
+        addr: &Address,
+        index: Option<usize>,
+    ) -> HistorySearchResult<Amount> {
+        let iter = self.0.iter().skip(index.unwrap_or_default()).rev();
+
+        for output in iter {
+            match output.state_changes.ledger_changes.0.get(addr) {
+                Some(SetUpdateOrDelete::Set(v)) => {
+                    return HistorySearchResult::Found(v.parallel_balance)
+                }
+                Some(SetUpdateOrDelete::Update(LedgerEntryUpdate {
+                    parallel_balance: SetOrKeep::Set(v),
+                    ..
+                })) => return HistorySearchResult::Found(*v),
+                Some(SetUpdateOrDelete::Delete) => return HistorySearchResult::Deleted,
+                _ => (),
```

Comment:
But this would change the behaviour. We are looking for an entry in the speculative ledger, that's why we return in the loop, to cut it if data was found. If data is not found we check every other slot in the history and `NotFound` only happens when it's been fully traversed without result.

Reply:
oh yes you're right, sorry